ironmon_tracker/Theme.lua

```diff
@@ -361,38 +446,160 @@ end
 	local presetDropdown = forms.dropdown(presetsForm, {["Init"]="Loading Presets"}, 50, 30, 145, 30)
 	forms.setdropdownitems(presetDropdown, Theme.PresetsOrdered, false) -- Required to prevent alphabetizing the list
 	forms.setproperty(presetDropdown, "AutoCompleteSource", "ListItems")
 	forms.setproperty(presetDropdown, "AutoCompleteMode", "Append")
 
-	forms.button(presetsForm, "Load", function()
-		Theme.importThemeFromText(Theme.PresetStrings[forms.gettext(presetDropdown)], true)
+	forms.button(presetsForm, "Preview", function()
+		local themeName = forms.gettext(presetDropdown)
+
+		for index, name in ipairs(Theme.PresetsOrdered) do
+			if name == themeName then
+				Theme.Screen.currentPreview = index
+				break
+			end
+		end
+
+		local themeCode = Theme.PresetStrings[Theme.PresetsOrdered[Theme.Screen.currentPreview or 1]]
+		Theme.Buttons.ApplyOrLoadTheme:updateState()
+		Theme.Buttons.RemoveTheme:resetButtonToDefault()
+		Theme.importThemeFromText(themeCode, false)
+
 		client.unpause()
 		forms.destroy(presetsForm)
 	end, 212, 29)
 end
 
+function Theme.openSaveCurrentThemeWindow()
+	Program.destroyActiveForm()
+	local form = forms.newform(350, 145, "Save Theme As...", function() client.unpause() end)
+	Program.activeFormId = form
+	Utils.setFormLocation(form, 100, 50)
+
+	forms.label(form, "Enter a name for this Theme:", 18, 10, 330, 20)
+	local saveTextBox = forms.textbox(form, "", 290, 30, nil, 20, 30)
+	forms.setproperty(saveTextBox, "MaxLength", 80)
+
+	Theme.Manager.SaveNewWarning = forms.label(form, "", 18, 55, 330, 20)
+	forms.setproperty(Theme.Manager.SaveNewWarning, "ForeColor", "Red")
+
+	Theme.Manager.SaveNewConfirm = forms.button(form, "Save", function()
+		-- Clear out warning texts
+		forms.settext(Theme.Manager.SaveNewWarning, "")
+
+		local formInput = forms.gettext(saveTextBox)
+
+		if formInput ~= nil and formInput ~= "" then
+			local success = false
+			local themeName = formInput
+
+			if themeName == Theme.PresetsOrdered[1] or themeName == Theme.PresetsOrdered[2] then
+				-- Don't allow importing "Current Theme (Custom)" or "Default Theme" as that is reserved
+				forms.settext(Theme.Manager.SaveNewWarning, "Cannot use a reserved Theme name")
+				forms.settext(Theme.Manager.SaveNewConfirm, "Save")
+			elseif themeName:find("%x%x%x%x%x%x") then
+				-- Don't allow six consectuive hexcode characters, as this screws with parsing it later
+				forms.settext(Theme.Manager.SaveNewWarning, "Name cannot have 6 consectuive hexcode characters (0-9A-F)")
+				forms.settext(Theme.Manager.SaveNewConfirm, "Save")
+			elseif Theme.PresetStrings[themeName] ~= nil then
+				-- If the Theme name is already in use, warn the user first
+				if forms.gettext(Theme.Manager.SaveNewConfirm) ~= "Confirm" then
+					forms.settext(Theme.Manager.SaveNewWarning, "A Theme with that name already exists. Overwrite?")
+					forms.settext(Theme.Manager.SaveNewConfirm, "Confirm")
+				else
+					success = true
+				end
+			end
```

Comment:
The code here only ever sets `success` to true when you press confirm on overwriting an old theme, meaning actual valid inputs of a theme name such as "Test" will not save with no warning messages or anything (the save button just does nothing)

Reply:
Good catch. I don't know why I wasn't just using `return` to escape out of the function when an issue arose. It should be fixed now.